### ISSUE TITLE
can't live watch logs when using loadblancer with different port than server

### ISSUE THREAD

**Issue by regevbr** [jhuckaby/Cronicle#185]
Leave the fact that it took me a few good hours to set up Cronicle using AWS ECS and put it behind an ELB (it was worth it though!), I came to realize that I can't watch live logs of job executions, since the web socket being open is using the port the server is running with (3012) even though it is only an internal port which is hidden behind a load balancer that uses port 80.

I guess the best solution is to modify the front end code here:

https://github.com/jhuckaby/Cronicle/blob/e13a3a896724094572b9624cd7e30fa0b00a651b/htdocs/js/pages/JobDetails.class.js#L867-L872

To be sort of the same as here:

https://github.com/jhuckaby/Cronicle/blob/e13a3a896724094572b9624cd7e30fa0b00a651b/htdocs/js/app.js#L263-L270

So it will respect the `web_direct_connect` configuration

I guess that can cause issues when using slaves etc, but currently I'm only using a single master only!

p.s.

I have read the other issues related to the live logs, but I'm simply propsing a quick fix so it can work when using a single master.

**Comment by BitProcessor** [comment]
Did you get this fixed ? Same issue here (on EKS behind LB)

**Comment by jhuckaby** [comment]
The fix proposed here will cause **many** other issues, so I cannot make the change as presented.  But I can say that this issue is entirely resolved in the next major version of Cronicle, where the live log system has been overhauled to share the main WebSocket and all data passes through the master server.  I don't have an ETA on this release however, as it is a major rewrite, and Cronicle is a one-man operation.

**Comment by BitProcessor** [comment]
Hi @jhuckaby - thanks for your reply. I'm running Cronicle in a single machine setup on kubernetes for some light tasks. I got the live logs working, by doing a simple dirty fix as:

```
	//	var url = app.proto + job.hostname + ':' + app.port;
	//	if (!config.web_socket_use_hostnames && app.servers && app.servers[job.hostname] && app.servers[job.hostname].ip) {
	//		// use ip if available, may work better in some setups
	//		url = app.proto + app.servers[job.hostname].ip + ':' + app.port;
	//	}

	  url = "https://cronicle.example.com"
```
where cronicle.example.com is the hostname pointing at the LB.
I would like to introduce an extra config variable such as "single_master_behind_lb" but I'm struggling to find on what place I need to introduce that extra variable, to be able to write the above in a more generic: if (single_master_behind_lb) url = <some_vars_to_get_to_the_same_result>.

Could you be so kind to point out to me:
- Which files to edit to introduce the new config variable and get it available in JobDetails.class.js
- What variable(s) I could use to get the correct value.

As for the next major version: could you also add a parameter to disable all local logging and spit out everything to stdout/stderr? That would be the docker-way :)

Thanks in advance for your help. Maarten

**Comment by jhuckaby** [comment]
@BitProcessor Very glad you found a fix.  The global `config` object is the correct place.  If you make a PR with this I will merge it.

Regarding your other request, you can do this right now.  In fact, I believe many people using Cronicle in Docker have a clean way to do it.  I think they have an official Docker image or something.  But anyway, I added them a special `--foreground` command-line argument for this very purpose, which launches Cronicle in the foreground (no daemon fork), and you can echo all logs to STDOUT using the `--echo` argument.  Meaning, put this in your Docker startup shell script:

```
node /opt/cronicle/lib/main.js --foreground --echo
```

There's no easy way to *disable* all local logging, but Cronicle rotates, archives and compresses its own logs every night at midnight (local server time), so they won't stack up.  Add a cron job to delete 30+ day old ones like this:

```
0 0 * * * find /opt/cronicle/logs/archives/ -type f -mtime 30 -delete
```

Good luck!

**Comment by jhuckaby** [comment]
You can also achieve this using these two [environment variables](https://github.com/jhuckaby/Cronicle#environment-variables), if that works better for you:

| Variable | Sample Value | Description |
|----------|--------------|-------------|
| `CRONICLE_foreground` | `1` | Run Cronicle in the foreground (no background daemon fork). |
| `CRONICLE_echo` | `1` | Echo the event log to the console (STDOUT), use in conjunction with `CRONICLE_foreground`. |

**Comment by BitProcessor** [comment]
Cool - I will update my docker image to use the main.js --foreground --echo command and add the log cleanup stuff. Thanks for your help!

PR with the websocket fix: https://github.com/jhuckaby/Cronicle/pull/226

(created a separate branch, because on master I also changed other stuff to allow the install script to work with my fork).

**Comment by BitProcessor** [comment]
Hmmz - seems like my fix doesn't actually work. It was working because of a previous dirty fix where I just set the url the desired value. That was cached in the browser. Seems like my extra config value is just empty, so I probably also need to change some other code to include that new config value?

**Comment by jhuckaby** [comment]
Huh, I honestly don't know.  I designed it so that anything you put into the `client` object in the main config.json file gets passed over to the browser and will be accessible in the global `config` object.  I have no idea why it isn't working for you.  Apologies 😞

**Comment by BitProcessor** [comment]
Ah! Does it have to be here: 
https://github.com/jhuckaby/Cronicle/blob/0d670ce9fdb9377834755a7a81861e501b69be41/sample_conf/config.json#L47 ?

**Comment by BitProcessor** [comment]
New PR: https://github.com/jhuckaby/Cronicle/pull/228

Successfully tested! 
Thanks for pointing out that my config value needed to be added in the client object.
The PR also adds fixes for the View and Download links while at the live logs screen.

**Comment by jhuckaby** [comment]
Released [v0.8.32](https://github.com/jhuckaby/Cronicle/releases/tag/v0.8.32) with this change in place.  Thanks @BitProcessor!
